src/shuffler-codereview.c

```diff
@@ -0,0 +1,1228 @@
+            if (!PokemonIsValidLevel(&tt->pool1[i], level)) {
+                totalPoolSize--;
+                continue;
+            }
+            filteredIndicies[nextIndex] = i;
+            nextIndex++;
+        }
+        // for pool2, loop through and record all the indicies of valid pokemon.
+        // we add the size of pool1 to our index because we're treating the
+        // three pools as a single pool. later when we actually pick a pokemon
+        // from the pool, we'll know that if the index is greater than pool1's
+        // size, it's in pool2 or pool3.
+        for (int i = 0; i < tt->pool2Size; i++) {
+            if (!PokemonIsValidLevel(&tt->pool2[i], level)) {
+                totalPoolSize--;
+                continue;
+            }
+            filteredIndicies[nextIndex] = i + tt->pool1Size;
+            nextIndex++;
+        }
+        // for pool3, loop through and record all the indicies of valid pokemon.
+        // again, we're adding the sizes of pools 1 and 2 to the index value.
+        for (int i = 0; i < tt->pool3Size; i++) {
+            if (!PokemonIsValidLevel(&tt->pool3[i], level)) {
+                totalPoolSize--;
+                continue;
+            }
+            filteredIndicies[nextIndex] = i + tt->pool1Size + tt->pool2Size;
+            nextIndex++;
+        }
+        // shuffle the indicies
+        for (int i = totalPoolSize - 1; i > 0; i--) {
+            int r = tinymt32_generate_uint32(&masterSeed) % (i + 1);
+            int tmp;
+            SWAP(filteredIndicies[i], filteredIndicies[r], tmp);
+        }
+        // take the top 6 values from the newly shuffled list of indicies.
+        for (int i = 0; i < 6; i++) {
+            partyIndicies[i] = filteredIndicies[i];
+        }
+        break;
+    }
+    case TB_STRAT_GYM_MEMBER: {
+        trainer->partySize = 4;
+        // figure out which pool is largest so we can allocate our array to fit all the indicies.
+        int maxPoolSize = tt->pool1Size;
+        if (maxPoolSize < tt->pool2Size) {
+            maxPoolSize = tt->pool2Size;
+        }
+        if (maxPoolSize < tt->pool3Size) {
+            // I don't think this should ever happen; pool3 should necessarily
+            // be smaller than pools 1 and 2. But just in case...
+            maxPoolSize = tt->pool3Size;
+        }
+
+        int filteredPoolSize = tt->pool1Size;
+        int filteredIndicies[maxPoolSize];
+        int nextIndex = 0;
+        // we start with pool1. we want to pick 3 pokemon total from pool1.
+        // filter out possible pokemon that are of the wrong level.
+        for (int i = 0; i < tt->pool1Size; i++) {
+            if (!PokemonIsValidLevel(&tt->pool1[i], level)) {
+                filteredPoolSize--;
+                continue;
+            }
+            filteredIndicies[nextIndex] = i;
+            nextIndex++;
+        }
+        // shuffle the remaining indicies.
+        for (int i = filteredPoolSize - 1; i > 0; i--) {
```

Comment:
I believe this is the fourth time I've read this code. Make it a function, and stop killing babies.